### ISSUE TITLE
Kafka ingestion inconsistency?

### ISSUE THREAD

**Issue by jianzzz** [ClickHouse/ClickHouse#52313]
Clickhouse Version: 22.8.5.1

I am using ClickHouse Kafka  engine to ingest data(using MATERIALIZED VIEW), data is continuously written in chronological order and partitioned by day, and I noticed a data skew.
So I adjusted the kafka_num_consumers parameter, executed the following sql:
```
detach view xxx on cluster default_cluster;
drop table xxx on cluster default_cluster;   # Kafka source table
create table xxx on cluster default_cluster;  # Kafka source table，this time I adjusted the kafka_num_consumers parameter
attach view xxx on cluster default_cluster;
```

I used another table to ingest data, and other parameters were consistent except that the consumption group was different.
Then I found the data inconsistency. I counted the totals for a given day(the partition that data was being written to before the change) and found they were inconsistent.

For comparison, I kept the parameters the same and repeated the above steps.

Before the change, data was being written to a partition (partition by day). After the data was written to the partition in that day, I still found that the data in the partition was inconsistent.

At the same time, I also counted the total data of the partition for the next day, which was consistent.

**Comment by jianzzz** [comment]
Is it because ClickHouse has no transaction guarantee and Kafka consumer offset is committed asynchronously, which can lead to repeated consumption of data

**Comment by morsapaes** [comment]
> Is it because ClickHouse has no transaction guarantee and Kafka consumer offset is committed asynchronously, which can lead to repeated consumption of data?

Exactly. The behavior you observed is expected given that the Kafka Table Engine provides _at-least-once_ semantics. Changing the `kafka_num_consumers` triggers a consumer group rebalancing, which can cause previously uncommitted offsets to be re-ingested. For that reason, and as suggested in the documentation, you should account for duplicates when using this table engine.

We've been working on an improved implementation of the Kafka Table Engine that uses Keeper to store offsets, rather than Kafka itself. This version will (eventually) provide stronger guarantees by supporting atomic commits.
